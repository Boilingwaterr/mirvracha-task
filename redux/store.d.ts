export declare const setInputValue: (inputValue: string) => {
    type: string;
    inputValue: string;
};
export declare const setFetching: (isFetching: boolean) => {
    type: string;
    isFetching: boolean;
};
export declare const setSize: (value: number) => {
    type: string;
    value: number;
};
export declare const setCountries: (list: []) => {
    type: string;
    list: [];
};
export declare const getAllCountries: () => (dispatch: Function) => Promise<void>;
export declare const getCountriesByName: (value: string) => (dispatch: Function) => Promise<void>;
export declare const store: import("redux").Store<import("redux").CombinedState<{
    countries: {
        inputValue: string;
        list: any;
        listSize: number;
        error: string;
        isFetching: boolean;
    };
}>, import("redux").AnyAction> & {
    dispatch: unknown;
};
