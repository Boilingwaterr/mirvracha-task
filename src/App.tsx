import React, { useEffect } from "react";
import { connect } from "react-redux";
import { SearchField } from "./components/SearchField";
import { MainContent } from "./components/MainContent";
import {
  getCountriesByName,
  getAllCountries,
  setSize,
  setInputValue,
} from "./redux/store";
import { Preloader } from "./components/Preloader";
import s from "./App.module.css";

const { app, list_container } = s;

interface Props {
  list: [];
  error: string;
  listSize: number;
  isFetching: boolean;
  getAllCountries: Function;
  getCountriesByName: Function;
  setSize: Function;
  setInputValue: Function;
  inputValue: string;
}

interface State {
  countries: {
    list: [];
    error: string;
    listSize: number;
    isFetching: boolean;
    inputValue: string;
  };
}

const AppComponent: React.FC<Props> = ({
  list,
  listSize,
  isFetching,
  error,
  getAllCountries,
  getCountriesByName,
  setSize,
  setInputValue,
  inputValue,
}) => {
  useEffect(() => {
    getAllCountries();
  }, [getAllCountries]);

  return (
    <div className={app}>
      <SearchField
        getCountriesByName={getCountriesByName}
        setSize={setSize}
        setInputValue={setInputValue}
        inputValue={inputValue}
      />
      {!!error ? (
        <h1>{error}</h1>
      ) : (
        (isFetching && <Preloader />) ||
        (list && (
          <div className={list_container}>
            <MainContent list={list} listSize={listSize} />
          </div>
        ))
      )}
    </div>
  );
};

const mapStateToProps = (state: State) => {
  return {
    list: state.countries.list,
    error: state.countries.error,
    listSize: state.countries.listSize,
    isFetching: state.countries.isFetching,
    inputValue: state.countries.inputValue,
  };
};

export const App = connect(mapStateToProps, {
  getAllCountries,
  getCountriesByName,
  setSize,
  setInputValue,
})(AppComponent);
