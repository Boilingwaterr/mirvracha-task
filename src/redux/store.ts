import { api } from "./../api/api";
import { createStore, combineReducers, applyMiddleware, compose } from "redux";
import thunkMiddleware from "redux-thunk";

const SET_COUNTRIES = "SET_COUNTRIES";
const SET_SIZE = "SET_SIZE";
const SET_ERROR = "SET_ERROR";
const SET_FETCHING = "SET_FETCHING";
const SET_INPUT_VALUE = "SET_INPUT_VALUE";

let initialState = {
  inputValue: "",
  list: null,
  listSize: 50,
  error: "",
  isFetching: false,
};

interface Action {
  inputValue: string;
  type: string;
  list?: [];
  error?: string;
  value?: number;
  isFetching: boolean;
}

const countries = (state = initialState, action: Action) => {
  switch (action.type) {
    case SET_COUNTRIES:
      return {
        ...state,
        list: action.list,
      };

    case SET_ERROR:
      return {
        ...state,
        error: action.error,
      };
    case SET_SIZE:
      return {
        ...state,
        listSize: action.value,
      };

    case SET_FETCHING:
      return {
        ...state,
        isFetching: action.isFetching,
      };

    case SET_INPUT_VALUE:
      return {
        ...state,
        inputValue: action.inputValue,
      };

    default:
      return state;
  }
};

export const setInputValue = (inputValue: string) => {
  return {
    type: SET_INPUT_VALUE,
    inputValue,
  };
};

export const setFetching = (isFetching: boolean) => {
  return {
    type: SET_FETCHING,
    isFetching,
  };
};

export const setSize = (value: number) => {
  return {
    type: SET_SIZE,
    value,
  };
};

const setError = (error: string) => {
  return {
    type: SET_ERROR,
    error,
  };
};

export const setCountries = (list: []) => {
  return {
    type: SET_COUNTRIES,
    list,
  };
};

export const getAllCountries = () => {
  return async (dispatch: Function) => {
    dispatch(setFetching(true));
    const result = await api.getAllCountries();
    dispatch(setCountries(result));
    dispatch(setFetching(false));
  };
};

export const getCountriesByName = (value: string) => {
  return async (dispatch: Function) => {
    if (!!value.trim()) {
      dispatch(setFetching(true));
      const result = await api.getCountryByName(value);
      if (result.status === 404) {
        dispatch(setError(result.message));
        dispatch(setFetching(false));
      } else {
        dispatch(setError(""));
        dispatch(setCountries(result));
        dispatch(setFetching(false));
      }
    } else {
      getAllCountries();
    }
  };
};

let reducers = combineReducers({
  countries,
});

export const store = createStore(
  reducers,
  compose(applyMiddleware(thunkMiddleware))
);
