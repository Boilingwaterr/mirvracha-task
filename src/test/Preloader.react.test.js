import React from "react";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import { App } from "../App";
import { store, setFetching } from "../redux/store";

test("render Preloader", () => {
  store.dispatch(setFetching(true));
  const component = renderer.create(
    <Provider store={store}>
      <App />
    </Provider>
  );

  expect(component).toMatchSnapshot(); //expecting to snapshot <Preloader/>
});
