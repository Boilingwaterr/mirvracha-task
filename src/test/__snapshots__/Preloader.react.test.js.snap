// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders countries 1`] = `
<div
  className="app"
>
  <form
    className="search_form"
  >
    <input
      className="search_field"
      onChange={[Function]}
      type="text"
      value=""
    />
    <div
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="drop_down"
      >
        list size
      </div>
      <div />
    </div>
  </form>
  <div
    className="preloader"
  >
    <div
      className="preloaderContent"
    >
      <div
        className="preloaderFirst"
      />
      <div
        className="preloaderSecond"
      />
      <div
        className="preloaderThird"
      />
    </div>
  </div>
</div>
`;
